File: java/group2/RotateArray.java
// Rotate a given array to the right by k steps, where k is non-negative.
// Sample input: Input: [5,15,25,35,45,55,65], k = 2
// Sample output: [55,65,5,15,25,35,45]
// Solution: time complexity: O(n), space complexity: O(1)

class RotateArray {
    public static void main(String[] args) {
        int[] myArray = {5,15,25,35,45,55,65};
        int k = 2;
        rotateArray(myArray, k);
        for (int i = 0; i < myArray.length; i++) {

            System.out.print("," + myArray[i]);
        }
        System.out.println();
    }

    private static void rotateArray(int[] myArray, int k) {
        k %= myArray.length;
        reverse(myArray, 0, myArray.length - 1);
        reverse(myArray, 0, k - 1);
        reverse(myArray, k, myArray.length - 1);
    }
    private static void reverse(int[] myArray, int start, int end) {
        while (start < end) {
            int temp = myArray[start];
            myArray[start] = myArray[end];
            myArray[end] = temp;
            start++;
            end--;
        }
    }

}

File: bytopic/array/python/group2/rotate_array.py
"""
Given an integer array nums, rotate the array to the right by k steps, where k is non-negative.

Example:
Input: nums = [1,2,3,4,5,6,7], k = 3
Output: [5,6,7,1,2,3,4]
Explanation:
rotate 1 steps to the right: [7,1,2,3,4,5,6]
rotate 2 steps to the right: [6,7,1,2,3,4,5]
rotate 3 steps to the right: [5,6,7,1,2,3,4]

Tag: 6/150
Tag: 189/2927, R54/2936 (overall frequency ranking)
"""


def rotate(nums: list[int], k: int) -> None:
    length = len(nums)
    if length == k:
        return
    k = k % length  # the case when k > length
    nums.reverse()
    for i in range(k//2):
        nums[i], nums[k-1-i] = nums[k-1-i], nums[i]
    for i in range(k, (length+k)//2):
        nums[i], nums[length-1-i+k] = nums[length-1-i+k], nums[i]


def main():
    nums = [1, 2, 3, 4, 5, 6, 7]
    k = 3
    rotate(nums, k)
    print(nums)


if __name__ == "__main__":
    main()
